src/crafting.cpp

```diff
@@ -922,43 +922,60 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
         cata::optional<time_duration> max_experience;
     };
 
-    // The proficiency, and the multiplier on the time we learn it for
-    std::vector<learn_subject> subjects;
-    for( const recipe_proficiency &prof : making.proficiencies ) {
-        if( !_proficiencies->has_learned( prof.id ) &&
-            prof.id->can_learn() &&
-            _proficiencies->has_prereqs( prof.id ) ) {
-            learn_subject subject{
-                prof.id,
-                prof.learning_time_mult / prof.time_multiplier,
-                prof.max_experience
-            };
-            subjects.push_back( subject );
+    const std::vector<Character *> all_crafters = [&]() {
+        std::vector<Character *> v;
+        const std::vector<npc *> helpers = get_crafting_helpers();
+        v.reserve( 1 + helpers.size() );
+        v.push_back( this );
+        v.insert( v.end(), helpers.begin(), helpers.end() );
+        return v;
+    }
+    ();
+
+    const auto practice_proficiencies = [&]( Character & p ) {
```

Comment:
Also too much lambda here, this would be better served by simply building a vector of Character * including the actor and helpers and looping over it.

Reply:
The function needs to return whether the player learned a proficiency, that would be hard to do by looping over a single vector.

Reply:
Not particularly difficult, either prepend `this` to the vector and only capture the value once, or append this to the end and overwrite it on each iteration, substituting `success = x; continue;` for `return x;`

Reply:
Done